### ISSUE TITLE
TestStoreZoneUpdateAndRangeSplit is flaky

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#3219]
The following test appears to have failed:

[#9587](https://circleci.com/gh/cockroachdb/cockroach/9587):

```
W1124 18:54:16.485918 903 storage/store.go:1341  conflicting intents on [/System/Meta2/"\xff\f\x03\xe8ZJfGAbTYxaPiwjCPvCNncIHEEbqsYLsbKTRkzCjHgsQETXLSnMYmnLAnuDesrEhwiTXIIKwtDgamIfhAewgELLRRZutEDtwxmMVK"]: resolved? true
I1124 18:54:16.493725 903 storage/replica_command.go:977  pushing intent from txn with lower priority "storage/replica_command.go:1318 (*Replica).AdminSplit" id=85c0326b key=/Local/Range/RangeDescriptor/"\xff\f\x03\xe8ZJfGAbTYxaPiwjCPvCNncIHEEbqsYLsbKTRkzCjHgsQETXLSnMYmnLAnuDesrEhwiTXIIKwtDgamIfhAewgELLRRZutEDtwxmMVK" rw=false pri=72.17366773 iso=SERIALIZABLE stat=PENDING epo=0 ts=0.000000000,766 orig=0.000000000,766 max=0.000000000,766 vs 1975588288
W1124 18:54:16.496811 903 storage/store.go:1341  conflicting intents on [/Local/Range/RangeDescriptor/"\xff\f\x03\xe8ZJfGAbTYxaPiwjCPvCNncIHEEbqsYLsbKTRkzCjHgsQETXLSnMYmnLAnuDesrEhwiTXIIKwtDgamIfhAewgELLRRZutEDtwxmMVK"]: resolved? true
W1124 18:54:16.503140 903 storage/store.go:1341  conflicting intents on [/Local/Range/RangeDescriptor/"\xff\f\x03\xe8"]: resolved? true
I1124 18:54:16.715820 903 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- FAIL: TestStoreZoneUpdateAndRangeSplit (1.50s)
	client_split_test.go:492: expected range to split within 1s
	client_split_test.go:497: second range has split: range_id:2 start_key:"\377\014\003\350" end_key:"\377\014\003\350ZJfGAbTYxaPiwjCPvCNncIHEEbqsYLsbKTRkzCjHgsQETXLSnMYmnLAnuDesrEhwiTXIIKwtDgamIfhAewgELLRRZutEDtwxmMVK" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
=== RUN   TestStoreRangeSplitWithMaxBytesUpdate
I1124 18:54:16.717557 903 gossip/gossip.go:172  gossiping node descriptor node_id:1 address:<network_field:"localhost" address_field:"1" > attrs:<> 
I1124 18:54:16.722214 903 multiraft/multiraft.go:557  node 1 starting
I1124 18:54:16.723475 903 raft/raft.go:409  [group 1] 1 became follower at term 5
I1124 18:54:16.723700 903 raft/raft.go:218  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I1124 18:54:16.723983 903 multiraft/multiraft.go:959  node 1 campaigning because initial confstate is [1]
I1124 18:54:16.724504 903 raft/raft.go:489  [group 1] 1 is starting a new election at term 5
I1124 18:54:16.724615 903 raft/raft.go:422  [group 1] 1 became candidate at term 6
--
I1124 18:54:22.780317 903 storage/replica.go:1369  gossiping system config from store 1, range 2
I1124 18:54:22.792994 903 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestStoreSplitReadRace (0.06s)
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.72s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	32.900s
=== RUN   TestBatchBasics
I1124 18:54:05.871518 962 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I1124 18:54:05.875567 962 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.01s)
=== RUN   TestBatchMerge
I1124 18:54:05.882461 962 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.01s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

**Comment by tschottdorf** [comment]
re #3301, @bdarnell are you fixing up this test (if so, assign yourself)? I haven't looked into it, but if the test failure is removed with #3301 then something's going wrong in this test.

**Comment by bdarnell** [comment]
To reproduce: `RANDOM_SEED=-1572944187333899517 make stressrace PKG=./storage TESTS=StoreZoneUpd`. It's not deterministic even with the fixed seed due to timing considerations.


This is a bug in `DistSender` in which write requests get separated from their `BeginTransaction`, resulting in intents that appear to belong to a transaction that has been GC'd when in fact it's a transaction that hasn't been written yet. This situation is especially likely for splits, but can occur with any multi-range transaction.

`AdminSplit` creates a transaction whose first batch looks like [BeginTransaction(rangeDescriptorKey), CPut(rangeDescriptorKey), Put(metaKey)]. If the rangeDescriptorKey and metaKey are on different ranges, DistSender must split this into two batches: [nil, nil, Put(metaKey)] and [BeginTransaction, CPut, nil]. It then executes these split batches *in key order*, writing the intent to the meta key before calling BeginTransaction. 

In between the two batches, it may have to do a RangeLookup, which will encounter the intent we just wrote. The RangeLookup does an inconsistent read, so it will lazily try to resolve any intents it encounters. If this happens before the BeginTransaction, it will see no transaction record and resolve the intent as aborted. This also poisons the sequence cache, causing the transaction to abort on its next request (but the transaction record will be left behind in the pending state, because everything assumes that if we've made it this far the transaction record is already marked as aborted).

The whole process will eventually retry, but since there is still a "pending" transaction record on disk, the retry can only succeed if it has a high enough priority to push the older abandoned transaction (or after the transaction has expired, but the test doesn't run long enough for that).

tl;dr: DistSender needs to be careful to preserve the invariant that BeginTransaction is executed before any other writes.

**Comment by bdarnell** [comment]
#3301 doesn't fix this test; it makes the failure easier to reproduce.

**Comment by tschottdorf** [comment]
ok. I'm milling about `DistSender` anyway; if you like, I can pick this up.

**Comment by tschottdorf** [comment]
Is there really a `PENDING` transaction entry? The Push should write an `ABORTED` one, if anything. From the write-up it seems like `BeginTransaction` never executes successfully. What you're describing sounds more like a multi-range request failing when it has both a transaction entry and a pending intent, but without an EndTransaction call (due to #3303). This should cause the intents to remain there, with a PENDING transaction entry, but with the new incarnation of the transaction running into its own old intents which remain unresolved until a Push succeeds.

**Comment by bdarnell** [comment]
The `BeginTransaction` does execute successfully and leave a PENDING transaction record. The full sequence of events is:

1. AdminSplit starts its transaction with [BeginTransaction, CPut(rangeDescriptor), Put(meta)]
2. First part of batch executes [nil, nil, Put(meta)]. Leaves an intent on the meta record.
3. Outside the transaction, a range lookup hits the intent on the meta record, does a PushTxn(rangeDescriptor) which doesn't find any persisted transaction and returns ABORTED, then resolves the intent and poisons the sequence cache on the range containing `meta`.
4. Second part of the transaction batch [BeginTransaction, CPut, nil] executes on the range containing the range descriptor. It succeeds (because the sequence cache isn't poisoned on this range) and writes a pending txn record.
5. The transaction continues. The next request is Get(rangeTreeRoot), which (in this case) lives in the same range as `meta`. This returns TransactionAbortedError due to the poisoned sequence cache.
6. The transaction starts over. It's a TransactionAbortedError instead of a TransactionRetryError so it gets a new ID and random priority; it doesn't recognize that the PENDING transaction left behind is its own past incarnation. If the new transaction has a higher priority than the older one, it will push and succeed (because now that the cache is warm there will be no RangeLookup to mess things up). If not, back off and retry again.

If you're already working in DistSender then I'll leave this one for you and move on to the next split-related test failure.

**Comment by tschottdorf** [comment]
thanks @bdarnell, that checks out. The decision to not "chunk off" the `BeginTransaction` call like we do with `EndTransaction` was a deliberate one, but now that we see errors here it's clear that we need to do something.

Looking towards a future version of `DistSender` which (mostly) parallelizes multi-range queries, it would be good to keep avoid having to have a separate first call (but before we get deep in the the trenches here, that's probably the right thing to do for now).

Another way to look at this issue is that the real problem is that the `Push` doesn't write an `ABORTED` entry. This is a correctness issue since it usually indicates that someone has since removed some of our intents. The reason it doesn't write one is otherwise there's some circularity in garbage-collection: sequence cache entries cause a Push, and it's awkward if those recreate transaction entries. But those can just get a new push mode (`GC_TXN` instead of `CLEANUP_TXN`) so that handling skipped intents would recreate them. That would already get the test back to passing and I'm going to implement that to fix the correctness issue within.

Another angle is that the intents don't get resolved. This is an atomicity issue (and mostly attached to #3303): `DistSender` may write some intents, but only if a success is reported back to `TxnCoordSender` will they be removed when the client finds itself aborted. We should avoid the `10s` timeout whenever we can, and to keep the txn record returned to the client consistent, we need to do something here anyways. This would also get the test passing.

Another thing we should consider is pushing the txn (instead of aborting). This would have the serializable transaction restart, so the "lost intents" is not an issue. As an upside, this would make `SNAPSHOT` transactions more performant (in the case in which the `BeginTransaction` doesn't get serialized to the front).
